courses/management/commands/create_verified_enrollment.py

```diff
@@ -0,0 +1,161 @@
+"""
+from django.contrib.auth import get_user_model
+from django.contrib.contenttypes.models import ContentType
+from django.core.management.base import BaseCommand, CommandError
+from django.db import transaction
+
+from courses.api import create_run_enrollments
+from courses.models import CourseRun, PaidCourseRun
+from ecommerce.api import fulfill_completed_order
+from ecommerce.discounts import DiscountType
+from ecommerce.models import Order, PendingOrder, Product, Discount
+from openedx.constants import EDX_ENROLLMENT_VERIFIED_MODE
+from users.api import fetch_user
+
+User = get_user_model()
+
+
+class Command(BaseCommand):
+    """creates an enrollment for a course run"""
+
+    help = "Creates an enrollment for a course run"
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--user",
+            type=str,
+            help="The id, email, or username of the User",
+            required=True,
+        )
+        parser.add_argument(
+            "--run",
+            type=str,
+            help="The 'courseware_id' value for the CourseRun",
+            required=True,
+        )
+        parser.add_argument(
+            "--code", type=str, help="The enrollment code for the course", required=True
+        )
+        parser.add_argument(
+            "-k",
+            "--keep-failed-enrollments",
+            action="store_true",
+            dest="keep_failed_enrollments",
+            help="If provided, enrollment records will be kept even if edX enrollment fails",
+        )
+        super().add_arguments(parser)
+
+    def handle(self, *args, **options):
+        """Handle command execution"""
+        user = fetch_user(options["user"])
+
+        run = CourseRun.objects.filter(courseware_id=options["run"]).first()
+        if run is None:
+            raise CommandError(
+                "Could not find course run with courseware_id={}".format(options["run"])
+            )
+
+        product = Product.objects.filter(
+            object_id=run.id, content_type=ContentType.objects.get_for_model(CourseRun)
+        ).first()
+        if product is None:
+            raise CommandError(
+                "No product found for that course with courseware_id={}".format(
+                    options["run"]
+                )
+            )
+
+        if run.course.is_country_blocked(user):
+            raise CommandError(
+                "Enrollment is blocked of this course with courseware_id={} for user {}".format(
+                    options["run"], options["user"]
+                )
+            )
+
+        # PaidCourseRun should only contain fulfilled or review orders
+        # but in order to avoid false positive passing in order__state__in here
+        paid_course_run = PaidCourseRun.objects.filter(
+            user=user,
+            course_run=run,
+            order__state__in=[Order.STATE.FULFILLED, Order.STATE.REVIEW],
+        ).first()
```

Comment:
This is a duplicate code and we should avoid it, and it is also very less likely that this would behave differently in different contexts. What I think is we should convert this code into a function and use it here and `ecommerce/models.py` so that any future changes in this logic apply everywhere.

Reply:
My second suggestion here is a bit related to the sequence of the checks we have in place. We are checking for product and blocked country above however if a user already has an enrollment those checks would be running as extra.

How about we take this check above along with the passed discount that we have below? so that we only check for country blocked or product of the user doesn't have enrollment already and the discount code is also valid?